Add sphinx extension

lars-petter-hauge opened the pull request:
This is very much a WIP, more to just have a starting point to work on

markusdregi commented:
This is exciting 🚀

lars-petter-hauge commented:
This PR is now starting to look like something useful. The tests are not exhaustive (and to be honest, it's slightly difficult to make tests for all the components here, perhaps I should add some functional test and include the html output from one of the schema's).

The functionality is pretty basic, it doesn't add any context/item validator/transformer yet. But I think as a first iteration, this PR is sufficient - this quickly grows.

I would suggest that we consider adding more `MetaKeys`, like `Title` on the containers, and perhaps `Example` for the types - these haven't been that relevant yet, but for building documentation they are. Currently we're slicing the description.. `title = schema.get(MK.Description, "default_title")[:10]`. And for the examples it think it would look really nice if the configuration could be copy-pasted and used right away.

lars-petter-hauge commented on configsuite/extension/ext.py:
```diff
@@ -0,0 +1,206 @@
+# we are assuming several keywords.
+# I think we should provide proper user feedback if these are not included.
+def generate(schema, parent_section, parent_title=""):
+    title = schema.get(MK.Description, "default_title")[:10]
```
This right here is just silly and needs to be fixed

lars-petter-hauge replied:
Created: https://github.com/equinor/configsuite/issues/117

lars-petter-hauge commented on configsuite/extension/ext.py:
```diff
@@ -0,0 +1,206 @@
+    cs_header,
+)
+
+BASIC_EXAMPLE = {
```
What I would like to have is the possibility of having a configuration example in the documentation that can be copy pasted directly and used. How about allowing the user to specify a default configuration that is used as an example? Instead of adding another `Example` key.

And if it is not provided, we'll leave it with `string` and `integer` etc.

lars-petter-hauge replied:
Created: https://github.com/equinor/configsuite/issues/116

lars-petter-hauge commented on configsuite/extension/ext.py:
```diff
@@ -0,0 +1,206 @@
+
+    section_content += nodes.paragraph(text=schema.get(MK.Description, ""))
+
+    for config_key in [MK.ElementValidators, MK.ContextValidators]:
```
It isn't possible to distinguish between validators/transformers and even descriptions. Should have some sort of markdown to demonstrate what it is

lars-petter-hauge replied:
Created: https://github.com/equinor/configsuite/issues/115

lars-petter-hauge commented on configsuite/extension/ext.py:
```diff
@@ -0,0 +1,206 @@
+
+    for config_key in [MK.LayerTransformation, MK.Transformation]:
+        if config_key in schema:
+            section_content += nodes.paragraph(text=schema.get(config_key).msg)
```
Is the `msg` part sufficient? Could we even use anything else?

lars-petter-hauge replied:
For the user of configsuite, I don't think it's obvious that the `msg` added to the decorators is what will inform the user in the documentation about what is going on. Also, I feel this splits the information about what is going into the documentation quite significantly. All the descriptions are already within the schema dict - and then you would have to go to the function declaration in order to edit documentation regarding transformations/validations..

And those `msg` looks to be one-liners, while transformations and validations, especially context based ones, could probably need more than that

Let's think about it a bit at least..

lars-petter-hauge replied:
https://github.com/equinor/configsuite/issues/114

lars-petter-hauge commented on configsuite/extension/ext.py:
```diff
@@ -0,0 +1,206 @@
+    section_content += section_children
+
+    section += section_content
+    section["example"] = example
```
I am not sure how robust this implementation is, adding an attribute to the `section` and then in the node later on, verify if it exists. But this was the simplest way of providing a mean to extract the info and put it into html code blocks. Note that if we provide latex etc as well, we can't do any language specific stuff here.

lars-petter-hauge replied:
For now this will be the method. Further investigation did not reveal a better way to do it.

lars-petter-hauge commented on configsuite/extension/ext.py:
```diff
@@ -0,0 +1,206 @@
+    app.add_node(
+        cs_container,
+        html=(visit_container, depart_container),
+        latex=(visit_container, depart_container),
```
the latex and text can't have the same nodes as html. We either have to supply individual nodes for all builders, or only support html.

lars-petter-hauge replied:
This PR will only handle `html`, the others have been removed

lars-petter-hauge commented on tests/test_license_header.py:
```diff
@@ -44,7 +44,7 @@ class TestLicenseHeaders(unittest.TestCase):
 
     def setUp(self):
-        ignore_extensions = (".pyc",)
+        ignore_extensions = (".pyc", ".rst", ".css", ".js")
```
Just add the licence to the `.css` and `.js` as well perhaps?

lars-petter-hauge commented on tests/test_sphinx_extension.py:
```diff
@@ -0,0 +1,124 @@
+        with open(os.path.join(app.outdir, "index.html"), "r") as f:
+            html = f.read()
+
+        self.assertTrue("PetMeet" in html)
```
This really isn't much of a test. I thought about adding the html output and test against that instead, but then if anything changes in the builder and sphinx, we would need to update the test - we shouldn't depend on external libraries..

I'll see if it's possible to test the nodes added by the extension instead.

lars-petter-hauge replied:
Expanded slightly, but did not manage to test the nodes directly

lars-petter-hauge commented on configsuite/extension/ext.py:
```diff
@@ -0,0 +1,206 @@
+            raise self.error(str(e))
+
+        try:
+            schema = getattr(module, func_str)()
```
The `func_str` is called in order to return the dict with a config_suite schema. Should this be implemented in a different manner? Should we validate that the dict is a valid schema?

oysteoh replied:
At least i would use module and func_str as argument to generate and return a schema. Not very found of funcs mutating input and no return

lars-petter-hauge replied:
> At least i would use module and func_str as argument to generate and return a schema. Not very found of funcs mutating input and no return.

Not sure if I follow you here. There is not modifying any arguments here. The `func_str` is just a string with the name of the function that we expect would return the `schema`. Given that the `func_str` indeed is a function name within the `module`, it's just called in order to retrieve the `schema`.

I wasn't sure if this was an okay method to use at all though, it seems a bit hacky, but then again how can it not? We're pretty much parsing a configuration file on the format of sphinx extensions.

> Not sure of the validation - i think just validating it's a dict and throw a proper error if not would be nice.

I'll do that!

oysteoh replied:
> Not sure if I follow you here. There is not modifying any arguments here. The `func_str` is just a string with the name of the function that we expect would return the `schema`. Given that the `func_str` indeed is a function name within the `module`, it's just called in order to retrieve the `schema`.
I misunderstood it seems the functionality behind `generate(...)`.. thought that one modified and generated the schema - which it doesn't .

markusdregi commented:
May we see the documentation for the schema levels in the documentation powered by the extension? 🚀

lars-petter-hauge commented:
Not sure what you're asking for? If the `MK.Description` is added? That is added indeed.

markusdregi commented:
No, more whether we should build the documentation for the layer schema 

https://github.com/equinor/configsuite/blob/01f9c5f94636f3d46eb29244b890b3dcb6fcac10/configsuite/schema.py#L71

as part of our documentation using the extension. So that configsuite's documentation is fuelled by the extension as well...

lars-petter-hauge commented:
Ah, I see. Not sure how it will work, but it could! And I do like the idea that we in fact use our own tools (and this is taking it a step further really)

Should probably add more descriptions to the `_META_SCHEMA` then, but that's completely doable!

lars-petter-hauge commented:
Just tested it on the current schema, and I mean, it works, just a bit bear bone :P

markusdregi commented:
We don’t have to do it in this PR; just thought it could be an interesting exercise and potentially useful 😉

oysteoh commented on configsuite/extension/ext.py:
```diff
@@ -0,0 +1,206 @@
+
+# we are assuming several keywords.
+# I think we should provide proper user feedback if these are not included.
+def generate(schema, parent_section, parent_title=""):
```
would it be possible splitting up the generate func making separate generators and use those inside this one? like - 'generate_content_section', 'generate_children_section' and such.... i think this method grasps to much to keep thing simple and clear.

lars-petter-hauge replied:
It certainly would, and I agree

oysteoh commented on configsuite/extension/ext.py:
```diff
@@ -0,0 +1,206 @@
+
+
+class ConfigsuiteDocDirective(Directive):
+    has_content = True
```
this used anywhere?

oysteoh commented on configsuite/extension/ext_nodes.py:
```diff
@@ -0,0 +1,80 @@
+        example_doc = example_doc.replace("\n", "<br>")
+        example_doc = example_doc.replace("    ", "&nbsp;&nbsp;")
+        self.body.append(
+            "<div class='cs_content'>Example:<br><code>{}</code>\n</div>".format(
```
what is the `\n` doing inside here?

lars-petter-hauge replied:
Unclear, probably just my habit.. I've replaced the `\n` with `<br>`.

oysteoh commented:
overall very good :+1:  - would like to see some kind of refactoring in the generate func if possible - think it contains a bit too much.

lars-petter-hauge commented:
I've updated the issues concerned by the reviewer, and I also added a bit more specific error handling for the `ConfigSuiteDirective` class when trying to call the `func_str`.
